Implement Resource Importer

amagana3 opened the pull request:
### Overview
This PR implements the `terraform import resource.resource_name resource_id` functionality. This did not exist for `alks_iamrole` or `alks_iamtrustrole`. Thankfully, we implemented this for the LTK resources, so it follows the exact same strategy. As long as the resources have an ID defined, then the resource can be looked up using the ID and it'll _magically_ work. Tested this in our `examples/` directory! 

### Steps to test:
1. Create an ALKS role in the UI
2. Try to create the same role in TFP
3. Apply, see the error that it already exists.
4. Import the resource by using it's name as the ID.
5. Once imported, re-apply and all should work!

### Example

```
➜  examples git:(implement_import) ✗ tf apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # alks_iamrole.test_role will be created
  + resource "alks_iamrole" "test_role" {
      + arn                      = (known after apply)
      + enable_alks_access       = false
      + id                       = (known after apply)
      + include_default_policies = false
      + ip_arn                   = (known after apply)
      + name                     = "TEST-DELETE"
      + role_added_to_ip         = (known after apply)
      + type                     = "AWS CodeBuild"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

alks_iamrole.test_role: Creating...

Error: Error creating role: [vx7zpien] Role already exists with the same name: TEST-DELETE

  on alks.tf line 36, in resource "alks_iamrole" "test_role":
  36: resource "alks_iamrole" "test_role" {


➜  examples git:(implement_import) ✗ tf import alks_iamrole.test_role TEST-DELETE
alks_iamrole.test_role: Importing from ID "TEST-DELETE"...
alks_iamrole.test_role: Import prepared!
  Prepared alks_iamrole for import
alks_iamrole.test_role: Refreshing state... [id=TEST-DELETE]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.

➜  examples git:(implement_import) ✗ tf plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

alks_iamrole.test_role: Refreshing state... [id=TEST-DELETE]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # alks_iamrole.test_role must be replaced
-/+ resource "alks_iamrole" "test_role" {
      ~ arn                      = "arn:aws:iam::336692926357:role/acct-managed/TEST-DELETE" -> (known after apply)
        enable_alks_access       = false
      ~ id                       = "TEST-DELETE" -> (known after apply)
      + include_default_policies = false # forces replacement
      + ip_arn                   = (known after apply)
      + name                     = "TEST-DELETE" # forces replacement
      + role_added_to_ip         = (known after apply)
      + type                     = "AWS CodeBuild" # forces replacement
    }

Plan: 1 to add, 0 to change, 1 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.

➜  examples git:(implement_import) ✗ tf apply -auto-approve
alks_iamrole.test_role: Refreshing state... [id=TEST-DELETE]
alks_iamrole.test_role: Destroying... [id=TEST-DELETE]
alks_iamrole.test_role: Destruction complete after 1s
alks_iamrole.test_role: Creating...
alks_iamrole.test_role: Creation complete after 1s [id=TEST-DELETE]

Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
```

webbbarker reviewed (commented):
@amagana3 Looks like these changes still result in a `ForceReplace` after import.  Was that intended?  I'd imagine in most cases folks just want to import and move on with the existing resource.

amagana3 commented:
@webbbarker great point. So you're right the reason it does the `ForceNew` is because the way `alks_iamrole` and `alks_iamtrustrole` are implemented. Their attributes `name`, `type`, and `include_default_policies` are set with `ForceNew: true`, and that is why it requires a replacement. I'm not 100% sure why those attributes were marked that way cause if you look at LTK, the only thing that is forced is the ID which is the `iam_username` of the LTK which is expected behavior for an ID. What're your thoughts?

webbbarker commented:
For LTKs, the only thing the user can modify via the ALKS API is the username (hence the only thing that forces new).  For IAM Roles, changing the IAM Role name, the IAM Role Type, or attaching/detaching default policies all modify the underlying role and force the role to be replaced.  Is there a way to clone the data into state without triggering an update?  I know that TF has mechanisms to communicate when a resource is actually modified but it gets a little weird.

amagana3 commented:
@webbbarker Yeah we can do that, it'll require a custom `importer` function instead of the OOB one since the default one only expects one attribute change (the ID). I don't think it'll be that much more difficult, so I'll try that out and get that pushed in the next commit.

amagana3 commented:
### Update:

After working on trying to get the resource to not destroy + create on an import, it was deemed impossible until we modify our existing backend to support `include_default_policies` as well as the role `type`. Once tags are available this story can be re-worked on.

I've cleaned up some of the code, left the import as is, and I've also added a warning label to the imports to let people know it'll force-replace the resource on import.

webbbarker commented on resource_alks_iamrole.go:
```diff
@@ -167,13 +170,10 @@ func resourceAlksIamTrustRoleCreate(d *schema.ResourceData, meta interface{}) er
 	d.SetId(response.RoleName)
-	d.Set("arn", response.RoleArn)
-	d.Set("ip_arn", response.RoleIPArn)
-	d.Set("role_added_to_ip", response.RoleAddedToIP)
```
I think this value is _only_ available when the role is created and I believe we've historically stuff that in the TF state.

amagana3 replied:
Good catch @webbbarker , fixed that. Re-added it to both create IAM Trust Role and regular Role. 👍🏼

webbbarker reviewed (changes requested):
One comment.

webbbarker reviewed (approved):
LGTM 👍
